### PR TITLE
Issue 1281 warning on no write

### LINKED ISSUE #1281: Cypress should not require write access of root folder
Why do we check if the root of the project is writeable? 👇 i.e. the ` fs.accessAsync(projectRoot, fs.W_OK) `

https://github.com/cypress-io/cypress/blob/2b2b6d99a9f1bf232d9c7396b25390913d5f2b18/packages/server/lib/util/settings.coffee#L76-L93

cc:/ @zacblazic

### PR DESCRIPTION
**Pull request by mrmeku**
- Closes #1281

### User facing changelog
Cypress no longer requires write access to the projectRoot, it instead will display a warning.

### Additional details
- This allows for cypress inetegration with other tools such as Bazel
- It allows Cypress to be run in sandboxed or otherwise readonly fs environments

#### Before

Screen Shot 2020-05-05 at 10 35 11 AM

#### After

Screen Shot 2020-05-05 at 10 25 28 AM

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?

### CODE REVIEW

**Comment by mrmeku** [comment]
@jennifer-shehane Could I get a review on this? Its a continuation of a PR you were reviewing: https://github.com/cypress-io/cypress/pull/5584

File: packages/server/lib/errors.coffee
Comment (jennifer-shehane) [inline packages/server/lib/errors.coffee:954]: Just updating some of the warning message to make it clear on the 'why' this is necessary plus some rewording.

```suggestion
    when "FOLDER_NOT_ACCESSIBLE"
      """
      Folder #{arg1} is not writable.
      
      Writing to this directory is required by Cypress in order to store screenshots and videos.

      Enable write permissions to this directory to ensure screenshots and videos are stored. 
      
      If you don't require screenshots or videos to be stored you can safely ignore this warning.
      """
```

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]

**Comment by mrmeku** [comment]
@jennifer-shehane Merged your changes! Thanks for those. Is there anything else you'd like me to do before this is ready to be merged?

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
When you add snapshots, this generates a file of the stdout output of the test you wrote. So, you need to run `SNAPSHOT_UPDATE=1 yarn test-e2e 8_read_only` to generate this file for comparison later (I've gone ahead and done this and committed it). The idea is - if something breaks, we'll see the changes in the STDOUT. 

Since the test you wrote has randomized characters - this won't work for snapshot testing. Each time you run the test, the generated output will be different. We need the test to have the same output each time the test is run.

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (mrmeku) [reply packages/server/test/e2e/8_read_only_fs.ts:10]: Ah, that's the magic incantation! Thanks so much for helping me out with that. Updating shortly

Comment (mrmeku) [reply packages/server/test/e2e/8_read_only_fs.ts]: Done!

**Comment by mrmeku** [comment]
@jennifer-shehane Snapshots updated and ready to go

**Review by jennifer-shehane — DISMISSED** [review]
There's an issue with the yarn lockfile that was committed. It's resolving to the incorrect has-binary dep. The has-binary dep should *not* have updated in your committed lockfile.

**Comment by mrmeku** [comment]
@jennifer-shehane Sorry I didn't catch that when I was reviewing my code. It's resolved now

File: packages/server/lib/util/settings.js
Comment (flotwig) [inline packages/server/lib/util/settings.js:129]: See how line 119 uses an object to filter the `.catch` block. You should be able to do this in a similar fashion:

```js
.catch({ code: 'EACCES' }, () => {
	return errors.warning(...)
})
```

File: packages/server/package.json
Comment (flotwig) [inline packages/server/package.json:204]: revert this plz. you may want to turn on the preference to "end files with a newline" in your editor

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (flotwig) [inline packages/server/test/e2e/8_read_only_fs.ts:45]:
```suggestion
    expect(err).to.include({ code: 'EACCES'})
```

maybe?

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (flotwig) [inline packages/server/test/e2e/8_read_only_fs.ts:17]:
```suggestion
    // we cant use one defined by source code since we need to modify the file permissions on the dir and files
```

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (flotwig) [inline packages/server/test/e2e/8_read_only_fs.ts:28]: hmm, I don't think you actually need to do this project setup manually, check out my comment on `e2e.it` for a simpler way

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (flotwig) [inline packages/server/test/e2e/8_read_only_fs.ts:35]: whoa, i didn't know this `0o` prefix was a thing, i am only familiar with using octals via the leading 0 (like `0500`)

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (flotwig) [inline packages/server/test/e2e/8_read_only_fs.ts:54]: Test projects are automatically scaffolded in `beforeEach`, so you can do all of your `chmodr` stuff via `onRun` on one of the project dirs that already exists:

```js
  e2e.it('warns when unable to write to dir', {
    project: Fixtures.projectPath('e2e'), // copied from `test/support/fixtures/projects/e2e`
    // use a failing spec so we can see a snapshot of the screenshot failing to be written
    // in addition to the video failing to be written
    spec: 'simple_failing_spec.coffee',
    expectedExitCode: 0,
    snapshot: true,
    onRun: (exec) => {
      // ...do the `chmodr` stuff on `Fixtures.projectPath('e2e')` here
      // ...assert that EACCES is thrown here
      // now call `exec` to run the e2e test
      return exec()
      // you may need an `afterEach` call to clean up the permissions
  })
```

This way you don't need to do any of the manual project setup you have in your `beforeEach`.

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (flotwig) [inline packages/server/test/e2e/8_read_only_fs.ts:7]: this is a nice, robust test. you should rename the file to `8_read_only_fs_spec.ts` to match how other e2e tests are named.

File: packages/server/__snapshots__/8_read_only_fs.ts.js
Comment (flotwig) [inline packages/server/__snapshots__/8_read_only_fs.ts.js:60]: would also like to see a screenshot failure here... using a spec that fails should generate a screenshot

**Review by flotwig — CHANGES_REQUESTED** [review]

File: packages/server/lib/util/settings.js
Comment (mrmeku) [reply packages/server/lib/util/settings.js:129]: done! thanks for the advise

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (mrmeku) [reply packages/server/test/e2e/8_read_only_fs.ts]: Done! Thanks :)

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (mrmeku) [reply packages/server/test/e2e/8_read_only_fs.ts]: Yeah, you totally have to do it this way when running in strict mode
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Deprecated_octal

Learn something new about java script everyday

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (mrmeku) [reply packages/server/test/e2e/8_read_only_fs.ts]: Done!

File: packages/server/test/e2e/8_read_only_fs.ts
Comment (mrmeku) [reply packages/server/test/e2e/8_read_only_fs.ts]: Thanks for the tip, it was driving me crazy that I couldn't do cleanup stuff before the afterEach Fixture code ran. That's why I wasn't using the built in fixture stuff

**Comment by flotwig** [comment]
Nyah... the e2e tests are failing because CI runs as root, and root doesn't get a permission denied error. Let me see how we can work around this.

**Review by flotwig — DISMISSED** [review]
Nice, I did some cleanup and made it so that failure to create the `videos` directory will also not crash the test. Added a new e2e test stage so we can run the tests as non-root as well.

File: circle.yml
Comment (bahmutov) [inline circle.yml:573]: why change chunking to strings?

File: packages/server/lib/errors.coffee
Comment (bahmutov) [inline packages/server/lib/errors.coffee:947]: i find it a little confusing: error is not accessible, but the real error is not writable

**Comment by flotwig** [comment]
Note that this PR will not fix errors caused by failure to "scaffold" a new project with a "supportFile". Some work needs to be done to not scaffold supportFolder in `run` mode:

https://github.com/mrmeku/cypress/blob/b5e46dd3bc7de54b4f9234754e0228ab64a9294e/packages/server/lib/project.js#L574-L582

And with that, it will need to be made possible for individual scaffolding steps to fail and cause the default values for `config.supportFile/etc.` to change, otherwise Cypress will just end up looking for non-existent files if scaffolding fails.

File: packages/server/lib/util/settings.js
Comment (bahmutov) [inline packages/server/lib/util/settings.js:125]: read or write :)

File: circle.yml
Comment (flotwig) [reply circle.yml:573]: added "non_root" chunk, and circleci has strict types for command parameters, so need to force these to be strings or else circleci complains about supplying an integer to a string parameter

File: packages/server/test/support/fixtures/projects/read-only-project-root/cypress/plugins/index.js
Comment (bahmutov) [inline packages/server/test/support/fixtures/projects/read-only-project-root/cypress/plugins/index.js:17]: nice

**Review by bahmutov — DISMISSED** [review]
I think this is good fix

**Comment by jennifer-shehane** [comment]
@flotwig @bahmutov Can this PR be approved?

**Comment by jennifer-shehane** [comment]
@flotwig Are you saying that this PR will not fully close #1281?

**Comment by flotwig** [comment]
@jennifer-shehane it's good to go, just waiting for those dang contributor tests to pass :roll_eyes: 

It closes #1281 since it will allow the folks in that thread to use Cypress on a read-only FS. I'll open a more specific issue about scaffolding requiring write access once this is closed.

**Review by flotwig — APPROVED** [review]

**Comment by yezhengli-Mr9** [comment]
Still https://github.com/cypress-io/cypress/issues/1281
```node
See discussion and possible solutions at
npm ERR! https://github.com/cypress-io/cypress/issues/1281
npm ERR! 
npm ERR! ----------
npm ERR! 
npm ERR! Failed to access /root/.cache/Cypress:
npm ERR! 
npm ERR! EACCES: permission denied, mkdir '/root/.cache/Cypress'
```
although tried 
```
 sudo chown -R $USER:$GROUP ~/.npm
sudo chmod +x /root/.cache
```

**Comment by jennifer-shehane** [comment]
@yezhengli-Mr9 Please open a new issue if you're encountering an issue with Cypress filling out the template.

**Pull request merged**
